### ISSUE TITLE
ssl handshake failed when using unencrypted websocket connection 

### ISSUE THREAD

**Issue by edwardotis** [tristandunn/pusher-fake-example#3]
The following error repeats for exactly 44 seconds on page load, and then finally the websocket connection will succeed. After that communication works fine. I'm running from the background javascript page of a chrome extension on same localhost. localhost is 64bit Ubuntu 14.04 LTS with Chrome 49

I don't understand why ssl is coming in to play at all? I've tried to set every relevant configuration on pusher fake server and javascript client to be unencrypted.

[16219:16258:0312/154027:ERROR:ssl_client_socket_openssl.cc(1149)] handshake failed; returned -1, SSL error code 1, net 
_error -100                                                                                                             
[16219:16258:0312/154027:ERROR:ssl_client_socket_openssl.cc(1149)] handshake failed; returned -1, SSL error code 1, net 
_error -100
[16219:16219:0312/154027:INFO:CONSOLE(228)] "Pusher : Connecting : {"transport":"ws","url":"ws://localhost:50389/app/<hidden>?protocol=7&client=js&version=3.0.0&flash=false"}", source: chrome-extension://<hidden>/background.js (228) 

pusher-js-3.0.0.js
        var pusher = new Pusher(PUSHER_API_KEY, {
            wsHost: 'localhost',
            wsPort: 50389,
            httpHost: 'localhost',
            httpPort: 39381,
            encrypted: false,
            disableStats: true,

            enabledTransports: ["ws"],
            authEndpoint: PUSHER_AUTH_URL,
            auth: {
                headers: {
                    'Authorization': authToken
                }
            }
        });

Config in pusher.rb

#config.serve_static_files = ENV["RAILS_SERVE_STATIC_FILES"].present?

Pusher.app_id = ENV["PUSHER_APP_ID"]
Pusher.key    = ENV["PUSHER_API_KEY"]
Pusher.secret = ENV["PUSHER_API_SECRET"]

PUSHER_HTTP_HOST = ENV["PUSHER_HTTP_HOST"] # 39381
HTTP_PORT = ENV["PUSHER_HTTP_PORT"] #localhost
PUSHER_WS_PORT = ENV["PUSHER_WS_PORT"] # 50389
PUSHER_WS_HOST = ENV["PUSHER_WS_HOST"] #localhost

require "pusher-fake"

Thin::Logging.silent = false
PusherFake.configuration.verbose = true
PusherFake.configuration.web_options = { host: PUSHER_HTTP_HOST, port: HTTP_PORT }
PusherFake.configuration.socket_options = { host: PUSHER_WS_HOST, port: PUSHER_WS_PORT, secure: false }

require "pusher-fake/support/base"

---Log---
No logging after boot until the 44 second mark.

=> Booting Thin
=> Rails 4.2.5 application starting in development on http://0.0.0.0:3000
=> Run `rails server -h` for more startup options
=> Ctrl-C to shutdown server
PusherFake socket_options
{:host=>"localhost", :port=>"50389", :secure=>false}
PusherFake web_options
{:host=>"localhost", :port=>"39381"}
Thin web server (v1.6.4 codename Gob Bluth)
Maximum connections set to 1024
Listening on 0.0.0.0:3000, CTRL+C to stop

Gemfile in pusher-fake-example:
source "https://rubygems.org"

gem "pusher", "0.17.0"
gem "pusher-fake", "1.5.0"
gem "rails",  "4.2.5"
gem "thin", "1.6.4"
gem "eventmachine", "1.0.9.1"

Any ideas?

**Comment by tristandunn** [comment]
Have you tried connecting to it from a regular page instead of from the Chrome extension? That's the one thing standing out to me from looking over the details here. If it works there but not in the extension then I'm not quite sure what's up.

**Comment by edwardotis** [comment]
Hi Tristan,

 I finally got to the bottom of it.
TL;DR 

pusher-fake-example works fine. No bug in it.

Long Explanation:
My setup on localhost for dev:
- gunicorn/Django server that hosts main website and also the /pusher/auth endpoint.
- pusher-fake-example rails server hosting websocket server on port 50389
- chrome extension client running pusher-js.js  and communicating to the django and the pusher-fake servers.

By using chrome://net-internals, I was able to see that Chrome was talking to the gunicorn/django python process instead of the pusher-fake ruby on rails process when it opened a websocket to "ws://localhost:50389"

I had [HSTS](https://docs.djangoproject.com/en/1.9/ref/middleware/#http-strict-transport-security) (http-strict-transport-security) enabled on my django server, and that automatically routed the request to wss, which was not enabled on any of my dev servers. And this resulted in the chrome_debug.log showing: ERROR:ssl_client_socket_openssl.cc(1149)] handshake failed; returned -1, SSL error code 1. 

So,  there appears to be a bug in Chrome where the browser's first websocket request targeted for port 50389 on localhost is being incorrectly sent to the process running on localhost at port 8000.  Maybe Chrome is caching the port and/or host from a previous http request, instead of correctly using the host and port passed into window.WebSocket() 

Anyway, I just disabled HSTS on my dev environments, and that works around the issue.

Best,

Ed

Netstat:
tcp        0      0 127.0.0.1:50389         0.0.0.0:*               LISTEN      38359/rails     
tcp        0      0 127.0.0.1:39381         0.0.0.0:*               LISTEN      38359/rails 
....
tcp        0      0 127.0.0.1:8000          0.0.0.0:*               LISTEN      41759/python3.5 

Logs example:
--
{
params: {
load_flags: 18,
method: "GET",
priority: "LOWEST",
url: "ws://localhost:50389/app/<hidden>?protocol=7&client=js&version=3.0.0&flash=false"
},
phase: 1,
source: {
id: 28829,
type: 1
},
time: "30665109",
type: 94
},
{
params: {
reason: "HSTS"
},
phase: 0,
source: {
id: 28829,
type: 1
},
time: "30665109",
type: 101
},
{
params: {
headers: [
"HTTP/1.1 307 Internal Redirect",
"Location: wss://localhost:50389/app/<hidden>?protocol=7&client=js&version=3.0.0&flash=false",
"Non-Authoritative-Reason: HSTS",
"Access-Control-Allow-Origin: chrome-extension://<hidden>",
"Access-Control-Allow-Credentials: true"
]
},
phase: 0,
source: {
id: 28829,
type: 1
},
time: "30665109",
type: 102
},
{
phase: 1,
source: {
id: 28829,
type: 1
},
time: "30665110",
type: 96
},
{
phase: 2,
source: {
id: 28829,
type: 1
},
time: "30665110",
type: 96
},
{
params: {
location: "wss://localhost:50389/app/<hidden>?protocol=7&client=js&version=3.0.0&flash=false"
},
phase: 0,
source: {
id: 28829,
type: 1
},
time: "30665110",
type: 95
},

...

{
params: {
error_lib: 33,
error_reason: 100,
file: "../../net/socket/ssl_client_socket_openssl.cc",
line: 2032,
net_error: -100,
ssl_error: 1
},
